explorer/interpreter/type_checker.h

```diff
@@ -79,8 +79,12 @@ class TypeChecker {
                                    int impl_offset) const
       -> Nonnull<const Witness*>;
 
+  // Returns the trace stream, if one was provided.
+  auto trace_stream() const -> std::optional<Nonnull<llvm::raw_ostream*>> {
```

Comment:
This feels a bit odd to add as a public API: I don't really think that a trace stream fits with type checking APIs... Of course, now I'm looking at the other things made public here, and maybe this is no longer really a basic TypeChecker API, but has morphed into some mix of public type checking API and internal utility holder?

Reply:
Yeah, I wonder if we should really have a single monolithic `TypeChecker` class at all. I think a different design for passing around the arena and trace stream would likely be better, and should allow us to make a lot of things that are currently `TypeChecker` members be independent functions / classes.

Short-term, I agree that this doesn't make a lot of sense as public API, and having removed the code with the TODO from `SingleStepEqualityContext` I'm not sure we really need this at all any more, so... removed.